File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Disease.java
package de.thkoeln.inf.gpm.vgb.model;

import model.DiseaseDao;
import util.DbUtil;

import java.util.List;

public class Disease {
    private Long id;
    private Integer category;
    private String description;

    public static Disease createOrUpdate(Disease disease) {
        return DbUtil.INSTANCE.runInTransaction(() -> DiseaseDao.Companion.save(disease));
    }

    public static List<Disease> findAll() {
        return DbUtil.INSTANCE.runInTransaction(DiseaseDao.Companion::findAll);
    }

    public static Disease findById(Long diseaseId) {
        return DbUtil.INSTANCE.runInTransaction(() -> DiseaseDao.Companion.get(diseaseId).toDisease());
    }

    public static void delete(Long diseaseId) {
        DbUtil.INSTANCE.runInTransaction(() -> DiseaseDao.Companion.delete(diseaseId));
    }

    public Disease(Long id, Integer category, String description) {
        this.id = id;
        this.category = category;
        this.description = description;
    }

    public Disease(Integer category, String description) {
        this.category = category;
        this.description = description;
    }

    public Long getId() {
        return id;
    }

    public Integer getCategory() {
        return category;
    }

    public void setCategory(Integer category) {
        this.category = category;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }
}

File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Disease.kt
package de.thkoeln.inf.gpm.vgb.model

import org.jetbrains.exposed.dao.IntEntity
import org.jetbrains.exposed.dao.IntEntityClass
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.transactions.transaction


class Disease(
    val id: Int?,
    val category: Int,
    val description: String

)


class DiseaseDao(id: EntityID<Int>) : IntEntity(id) {
    private var category by Diseases.category
    private var description by Diseases.description

    companion object : IntEntityClass<DiseaseDao>(Diseases) {
        /**
         * Update or create [Disease] in database
         */
        fun save(disease: Disease): Disease? = transaction {
            val newDisease = if (disease.id == null) {
                new { update(disease) }
            } else {
                val old = DiseaseDao.findById(disease.id)
                old?.update(disease)
                DiseaseDao.findById(disease.id)
            }

            newDisease?.toDisease()
        }
    }

    private fun update(disease: Disease) {
        this.category = disease.category
        this.description = disease.description
    }

    fun toDisease() = Disease(id.value, category, description)
}


object Diseases : IntIdTable() {
    val category = integer("kategorie")
    val description = text("beschreibung")
}